File: test/sync-run.test.mjs
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { findNewBookmarks, fetchWorks, nextGap, MIN_GAP_MS } from '../app/core/sync/run.js';

/* The shape parsePagination actually returns. A fake that invents a field the
   parser does not produce tests nothing but itself. */
const page = (ids, total = 3) => ({
  works: ids.map((workId) => ({ workId })),
  pagination: { current: 1, total },
});

test('walking stops once a page holds nothing new', async () => {
  const asked = [];
  const held = new Set(['old1', 'old2']);
  const { workIds } = await findNewBookmarks({
    fetchPage: async (p) => { asked.push(p); return page(p === 1 ? ['new1', 'old1'] : ['old1', 'old2']); },
    isHeld: (id) => held.has(id),
  });
  /* Bookmarks are newest first, so a page with nothing new means the rest was
     gathered on an earlier run. A full walk is 86 pages; this is two. */
  assert.deepEqual(asked, [1, 2]);
  assert.deepEqual(workIds, ['new1']);
});

test('everything new across several pages is collected', async () => {
  const { workIds } = await findNewBookmarks({
    fetchPage: async (p) => page(p === 1 ? ['a', 'b'] : p === 2 ? ['c'] : ['d']),
    isHeld: () => false,
  });
  assert.deepEqual(workIds, ['a', 'b', 'c', 'd']);
});

test('walking stops at the last page the archive reports', async () => {
  let calls = 0;
  await findNewBookmarks({
    fetchPage: async () => { calls++; return page(['x' + calls], 2); },
    isHeld: () => false,
  });
  assert.equal(calls, 2, 'it does not ask for a page beyond the end');
});

test('a sync can be stopped while it is walking', async () => {
  let calls = 0;
  const { workIds } = await findNewBookmarks({
    fetchPage: async () => { calls++; return page(['y' + calls], 50); },
    isHeld: () => false,
    shouldStop: () => calls >= 2,
  });
  assert.equal(calls, 2);
  assert.equal(workIds.length, 2);
});

test('a work that cannot be fetched does not end the sync', async () => {
  const { added, failed } = await fetchWorks({
    workIds: ['1', '2', '3'],
    fetchWork: async (id) => {
      if (id === '2') throw new Error('deleted');
      return { workId: id };
    },
    wait: async () => {},
  });
  // a bookmark outlives the work it points at
  assert.deepEqual(added.map((a) => a.workId), ['1', '3']);
  assert.deepEqual(failed, [{ workId: '2', reason: 'deleted' }]);
});

test('it waits between works, but not before the first', async () => {
  const waits = [];
  await fetchWorks({
    workIds: ['1', '2', '3'],
    fetchWork: async (id) => ({ workId: id }),
    wait: async (ms) => { waits.push(ms); },
  });
  assert.equal(waits.length, 2, 'two gaps for three works');
  for (const ms of waits) assert.ok(ms > 0);
});

test('fetching stops when asked', async () => {
  let done = 0;
  await fetchWorks({
    workIds: ['1', '2', '3', '4'],
    fetchWork: async (id) => { done++; return { workId: id }; },
    wait: async () => {},
    shouldStop: () => done >= 2,
  });
  assert.equal(done, 2);
});

test('the gap between requests is varied, not a metronome', () => {
  const gaps = Array.from({ length: 400 }, () => nextGap());
  const unique = new Set(gaps).size;
  assert.ok(unique > 100, 'a fixed interval is the easiest thing to notice');

  for (const g of gaps) {
    assert.ok(g >= MIN_GAP_MS * 0.45, `${g} is too close to the last request`);
    assert.ok(g <= MIN_GAP_MS * 2.5, `${g} waits absurdly long`);
  }
  const mean = gaps.reduce((a, b) => a + b, 0) / gaps.length;
  assert.ok(mean > MIN_GAP_MS * 0.6 && mean < MIN_GAP_MS * 1.6, `mean gap ${Math.round(mean)}ms`);
});

/* ------------------------------------------------- an author's whole listing */
const { walkListing, listingCost, shouldWalkWholeListing, PER_LISTING_PAGE } =
  await import('../app/core/sync/run.js');

test('the size of a listing is known from its page count alone', () => {
  const cost = listingCost(3);
  assert.equal(cost.pages, 3);
  assert.equal(cost.works, 3 * PER_LISTING_PAGE);
  assert.ok(cost.minutes >= 1, 'and roughly how long it will take');
});

test('a small author opens without being asked; a prolific one asks first', () => {
  assert.equal(shouldWalkWholeListing(1), true, '20 works is a single request');
  assert.equal(shouldWalkWholeListing(9), true, '180 works is under the threshold');
  assert.equal(shouldWalkWholeListing(10), false, '200 is not fewer than 200');
  assert.equal(shouldWalkWholeListing(40), false, '800 works is minutes of waiting');
});

test('walking a listing keeps works already held', () => {
  // an author page is asked for to see all of it, not only the unfamiliar parts
  return walkListing({
    fetchPage: async (p) => page(p === 1 ? ['a', 'b'] : ['c'], 2),
    wait: async () => {},
  }).then(({ works, totalPages }) => {
    assert.deepEqual(works.map((w) => w.workId), ['a', 'b', 'c']);
    assert.equal(totalPages, 2);
  });
});

test('walking a listing can be stopped', async () => {
  let calls = 0;
  const { works } = await walkListing({
    fetchPage: async () => { calls++; return page(['w' + calls], 20); },
    wait: async () => {},
    shouldStop: () => calls >= 3,
  });
  assert.equal(works.length, 3);
});

/* ------------------------------------------------- worth trying again? */
const { isTransient, retryDelay } = await import('../app/core/sync/run.js');

test('a server having a bad moment is worth trying again', () => {
  for (const reason of [
    'The archive answered 500', 'The archive answered 503',
    'could not reach the archive', 'network error', 'timed out',
  ]) assert.equal(isTransient(reason), true, reason);
});

test('a work that is gone is not worth trying again', () => {
  /* Asking a second time for a deleted work is rude and pointless, and
     counting a 500 alongside it writes off something that was probably fine a
     minute later. */
  for (const reason of [
    'That work does not exist, or has been deleted',
    'The archive answered 404', 'The archive answered 403',
    'No chapters found. The work may be restricted, deleted, or need a login.',
  ]) assert.equal(isTransient(reason), false, reason);
});

test('waiting longer after each failure', () => {
  assert.ok(retryDelay(1) > retryDelay(0));
  assert.ok(retryDelay(2) > retryDelay(1));
  assert.equal(retryDelay(9), retryDelay(3), 'and it stops growing rather than waiting for ever');
});

/*
 * Never getting there is always worth trying again.
 *
 * A phone loses a host for a moment — a tunnel, a handover between masts, a
 * network that has not finished coming up — and every one of those arrived as
 * "UnknownHostException: Unable to resolve host", which matched nothing here
 * and so was treated as a verdict on the work. A few seconds without signal
 * permanently gave up on whatever was in flight, and the reader was told the
 * work had been skipped, which was neither true nor anything they could act on.
 */
test('failing to reach the archive is worth trying again', () => {
  for (const reason of [
    'The app could not reach the archive: UnknownHostException: Unable to resolve host'
      + ' "archiveofourown.org": No address associated with hostname',
    'UnknownHostException: Unable to resolve host',
    'java.net.SocketTimeoutException: timeout',
    'SSLException: Read error: ssl=0x0: I/O error during system call',
    'ConnectException: Connection refused',
    'unexpected end of stream',
    'the archive answered 502',
  ]) {
    assert.equal(isTransient(reason), true, `should be retried: ${reason.slice(0, 60)}`);
  }
});

test('the archive saying no is still not worth trying again', () => {
  for (const reason of [
    'the archive answered 404',
    'That work does not exist, or has been deleted',
    'That work is restricted — sign in to the archive first',
    'no chapters found',
  ]) {
    assert.equal(isTransient(reason), false, `should not be retried: ${reason}`);
  }
});

File: dart/folio_core/test/sync_run_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:folio_core/folio_core.dart';
import 'package:test/test.dart';

/// The walks, and what the app decides about a failure.
///
/// `tools/emit-sync-conformance.mjs` runs a body of failure reasons through
/// the 1.x implementation and records what it said. The order of those rules
/// is the whole answer — a 429 has to be read as "slow down" before the
/// blanket 4xx rule reads it as "no" — and getting it wrong writes off every
/// work in flight the moment the archive starts throttling.
void main() {
  final fixture =
      jsonDecode(File('test/conformance/sync.json').readAsStringSync())
          as Map<String, Object?>;

  group('worth trying again, or not', () {
    for (final entry
        in (fixture['failures'] as List).cast<Map<String, Object?>>()) {
      final reason = entry['reason'] as String?;
      test('"${reason ?? '(nothing said)'}"', () {
        expect(
          isTransient(reason),
          entry['transient'],
          reason: 'the two versions disagree about whether to retry this',
        );
      });
    }
  });

  test('and waits longer each time, up to a point', () {
    for (final entry
        in (fixture['retries'] as List).cast<Map<String, Object?>>()) {
      expect(
        retryDelay(entry['attempt']! as int).inMilliseconds,
        entry['ms'],
      );
    }
  });

  test('a listing costs what 1.x says it costs', () {
    expect(perListingPage, fixture['perListingPage']);
    for (final entry
        in (fixture['listings'] as List).cast<Map<String, Object?>>()) {
      final pages = entry['pages'] as int?;
      final cost = entry['cost']! as Map<String, Object?>;
      final actual = listingCost(pages);
      expect(actual.pages, cost['pages'], reason: 'pages for $pages');
      expect(actual.works, cost['works'], reason: 'works for $pages');
      expect(actual.minutes, cost['minutes'], reason: 'minutes for $pages');
      expect(shouldWalkWholeListing(pages), entry['whole']);
    }
  });

  group('walking the bookmarks', () {
    ListingPage page(List<String> ids, {int? total}) =>
        ListingPage(workIds: ids, totalPages: total);

    test('stops where the bookmarks stop being new', () async {
      /* Bookmarks are listed newest first, so a page where every one was
         already known means the rest was too. A full re-walk is 86 pages. */
      final asked = <int>[];
      final result = await findNewBookmarks(
        fetchPage: (p) async {
          asked.add(p);
          return page(p == 1 ? ['new1', 'old1'] : ['old2', 'old3'], total: 86);
        },
        isHeld: (id) => id.startsWith('old'),
        isBookmarked: (id) => id.startsWith('old'),
      );

      expect(asked, [1, 2], reason: 'page two was all familiar; stop there');
      expect(result.workIds, ['new1']);
      expect(result.seen, ['new1', 'old1', 'old2', 'old3']);
    });

    test('a work already held can still be a new bookmark', () async {
      /* The bug this is here for: "is this among my bookmarks" and "do I have
         its text" are different facts, and 1.x used the second for both. So a
         work you already had and bookmarked today recorded nothing, and a
         page of works held for other reasons ended the walk early. */
      var pages = 0;
      final result = await findNewBookmarks(
        fetchPage: (p) async {
          pages++;
          return page(p == 1 ? ['held'] : ['known'], total: 5);
        },
        isHeld: (_) => true,
        isBookmarked: (id) => id == 'known',
      );

      expect(pages, 2, reason: 'page one was a new bookmark, so keep going');
      expect(result.seen, ['held', 'known'],
          reason: 'bookmark state is written from everything walked');
      expect(result.workIds, isEmpty, reason: 'nothing needs downloading');
    });

    test('and stops when asked to', () async {
      var pages = 0;
      var stop = false;
      final result = await findNewBookmarks(
        fetchPage: (p) async {
          pages++;
          stop = true;
          return page(['a$p'], total: 40);
        },
        isHeld: (_) => false,
        shouldStop: () => stop,
      );
      expect(pages, 1);
      expect(result.workIds, ['a1']);
    });
  });

  group('walking a listing', () {
    test('sees all of it, familiar works included', () async {
      /* Unlike the bookmark sync, an author page is asked for in order to see
         all of it, so a work already held still belongs in the list. */
      final walk = await walkListing(
        fetchPage: (p) async => ListingPage(
          workIds: ['w${p}a', 'w${p}b'],
          totalPages: 3,
        ),
      );
      expect(walk.totalPages, 3);
      expect(walk.workIds, ['w1a', 'w1b', 'w2a', 'w2b', 'w3a', 'w3b']);
    });

    test('an empty page is the end of it', () async {
      final walk = await walkListing(
        fetchPage: (p) async =>
            ListingPage(workIds: p == 1 ? ['only'] : const []),
      );
      expect(walk.workIds, ['only']);
    });
  });

  group('fetching a list of works', () {
    test('one that cannot be had is stepped over, not fatal', () async {
      /* A bookmark outlives the work it points at, and one deleted story
         should not end a sync. */
      final result = await fetchWorks<String>(
        workIds: ['a', 'gone', 'b'],
        fetchWork: (id) async {
          if (id == 'gone') throw Exception('that work does not exist');
          return id.toUpperCase();
        },
      );
      expect(result.added, ['A', 'B']);
      expect(result.failed.single.workId, 'gone');
      expect(result.failed.single.transient, isFalse,
          reason: 'a work that is gone will be gone next time too');
    });

    test('and one that was throttled is worth queueing again', () async {
      final result = await fetchWorks<String>(
        workIds: ['a'],
        fetchWork: (_) async => throw Exception('the archive answered 429'),
      );
      expect(result.failed.single.transient, isTrue);
    });

    test('progress is reported per work, not per batch', () async {
      final seen = <int>[];
      await fetchWorks<String>(
        workIds: ['a', 'b', 'c'],
        fetchWork: (id) async => id,
        onProgress: (p) => seen.add(p.done!),
      );
      expect(seen, [1, 2, 3]);
    });
  });
}
